File: src/com/pytest_support/pytest_hooks/PytestHooksGoToDeclarationHandler.java
package com.pytest_support.pytest_hooks;

import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandlerBase;
import com.intellij.openapi.editor.Editor;
import com.intellij.psi.PsiElement;
import com.pytest_support.utils.PytestUtils;
import com.pytest_support.utils.TypeUtils;
import org.jetbrains.annotations.Nullable;

import java.util.Objects;

public class PytestHooksGoToDeclarationHandler extends GotoDeclarationHandlerBase {
    @Nullable
    @Override
    public PsiElement getGotoDeclarationTarget(@Nullable PsiElement psiElement, Editor editor) {
        if (psiElement != null && PytestUtils.isTestOrPytestFile(psiElement)) {
            psiElement = psiElement.getParent();
            String text = psiElement.getText();

            return TypeUtils.getModuleFunctions(psiElement.getProject(), "_pytest.hookspec")
                    .stream()
                    .filter(function -> Objects.equals(function.getName(), text))
                    .findFirst()
                    .orElse(null);
        }
        return null;
    }
}

File: src/com/pytest_support/pytest_hooks/PytestHooksGoToDeclarationHandler.kt
package com.pytest_support.pytest_hooks

import com.intellij.codeInsight.navigation.actions.GotoDeclarationHandlerBase
import com.intellij.openapi.editor.Editor
import com.intellij.psi.PsiElement
import com.pytest_support.extensions.isInsideTestFile
import com.pytest_support.extensions.moduleFunctions

class PytestHooksGoToDeclarationHandler : GotoDeclarationHandlerBase() {
    override fun getGotoDeclarationTarget(psiElement: PsiElement?, editor: Editor): PsiElement? {
        if (psiElement?.isInsideTestFile == true) {
            val text = psiElement.parent!!.text

            return psiElement
                    .project
                    .moduleFunctions("_pytest.hookspec")
                    .firstOrNull { it.name == text }
        }

        return null
    }
}
